### PR TITLE
SE: Learn ObjectConstraint on branching from Recursive pattern

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Contributes to #5970

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp9.cs [hunk 1]
```diff
@@ -76,14 +76,6 @@ public void PatternMatching(object arg)
...
             o.ToString();       // Noncompliant
         }
-        if (arg is not not null)
-        {
-            arg.ToString();     // Non-compliant FIXME arg is null here
-        }
-        if (!(arg is not null))
-        {
-            arg.ToString();     // Non-compliant FIXME arg is null here
-        }
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp9.cs:86]: Unrelated fix of UTs: here, `arg` is known to be not null because there is `arg.ToString()` above. So condition is always false and invocation is not visited.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Pattern.cs [hunk 2]
```diff
@@ -148,15 +161,22 @@ declaration switch
...
+        private static bool RecursivePatternAlwaysMatchesAnyNotNull(IRecursivePatternOperationWrapper recursive) =>
+            recursive switch
+            {
                 {
                     PropertySubpatterns.Length: 0,
                     DeconstructionSubpatterns.Length: 0,
-                } => BoolConstraint.True,
-                _ when SubPatternsAlwaysMatch(recursive) => BoolConstraint.True,
-                _ => null,
+                } => true,
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Pattern.cs:174]: The type compatibility check is needed here too. Add some test cases for this. This is just a shortcut to prevent the expensive SubPatternsAlwaysMatches call. I think it is better to move this check into SubPatternsAlwaysMatches itself.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Pattern.cs [hunk 3]
```diff
@@ -148,15 +161,22 @@ declaration switch
             recursive switch
             {
                 _ when valueConstraint == ObjectConstraint.Null => BoolConstraint.False,
+                _ when RecursivePatternAlwaysMatchesAnyNotNull(recursive) => BoolConstraint.True,
+                _ => null,
+            };
+
+        private static bool RecursivePatternAlwaysMatchesAnyNotNull(IRecursivePatternOperationWrapper recursive) =>
+            recursive switch
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Pattern.cs:169]: Use `is` expression.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Pattern.cs [hunk 4]
```diff
@@ -148,15 +161,22 @@ declaration switch
...
+            {
                 {
                     PropertySubpatterns.Length: 0,
                     DeconstructionSubpatterns.Length: 0,
-                } => BoolConstraint.True,
-                _ when SubPatternsAlwaysMatch(recursive) => BoolConstraint.True,
-                _ => null,
+                } => true,
+                _ when recursive.InputType.Equals(recursive.MatchedType) && SubPatternsAlwaysMatches(recursive) => true,
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Pattern.cs:175]: Use DerivesOrImplements. There might also be the possibility to leverage "NarrowedType == MatchedType" or something like that instead.

Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Pattern.cs:175]: Add test cases for `someObj is string { }` and `someString is object { }`. Also: with non-empty patterns and some negations.

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs [hunk 5]
```diff
@@ -433,6 +433,59 @@ public void Branching_LearnsObjectConstraint_TypePattern_Negated(string expressi
...
 
+        [DataTestMethod]
+        [DataRow("arg is { }")]
+        [DataRow("arg is not not { }")]
+        [DataRow("!!(arg is { })")]
+        public void Branching_LearnsObjectConstraint_RecursivePattern_ElseIsNull(string expression)
+        {
+            var validator = CreateIfElseEndValidatorCS(expression, OperationKind.RecursivePattern);
+            validator.ValidateTag("If", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue());
+            validator.ValidateTag("Else", x => x.HasConstraint(ObjectConstraint.Null).Should().BeTrue());
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs:444]: Add tests were arg is a value type. There shouldn't be constraint here (there shouldn't be a constraint in if either). Add also tests with the other type kinds (e.g. unconstraint open generic).

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Pattern.cs (hunk 4 above)
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Pattern.cs:175]: I missed the type check in my PR. Can you add some tests to the "BoolConstraint" tests?

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs [hunk 6]
```diff
@@ -433,6 +433,59 @@ public void Branching_LearnsObjectConstraint_TypePattern_Negated(string expressi
...
+        [DataRow("arg is not string { Length: 0 }")]
+        public void Branching_LearnsObjectConstraint_RecursivePattern_Negated_IfIsUnknown(string expression)
+        {
+            var validator = CreateIfElseEndValidatorCS(expression, OperationKind.RecursivePattern);
+            validator.ValidateTag("If", x => x.Should().BeNull("it could be null or any other type"));
+            validator.ValidateTag("Else", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue());
+            validator.TagValues("End").Should().HaveCount(2)
+                .And.ContainSingle(x => x == null)
+                .And.ContainSingle(x => x != null && x.HasConstraint(ObjectConstraint.NotNull));
+        }
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs:487]: Add tests with deconstruction.

Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs:487]: What about tuples?
```cs
if ((a, b) is ({ }, { }))
{
  // a and b are not null
}
else
{
  // a or b or both are null
}
```

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp9.cs (hunk 1 above)
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp9.cs:86]: I did this because of https://github.com/SonarSource/sonar-dotnet/pull/6029#discussion_r958516524
Thanks for moving this to its own methods (I should have done this in the first place).

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp9.cs [hunk 7]
```diff
@@ -110,6 +102,46 @@ public void PatternMatching(object arg)
...
+            arg.ToString();     // Noncompliant
+        }
+    }
+
+    public void PatternMatching_NotNotNull_Mixed(object arg)
+    {
+        if (!(arg is not null))
+        {
+            arg.ToString();     // Noncompliant
+        }
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp9.cs:118]: Add else branches as in the other test cases

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs (hunk 5 above)
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs:444]: Hint: This is relevant for other patterns, like #6059 too.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp9.cs (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp9.cs:86]: It was not easy to spot

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Pattern.cs (hunk 2 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Pattern.cs:174]: The `SubPatternsAlwaysMatches` has a good name given what it does. I've prepended the type check, as it is needed always. That allows so simplify the syntax too

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Pattern.cs (hunk 3 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Pattern.cs:169]: Not sure if I like it, but it definitely removes some lines

Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Pattern.cs:169]: S3101 isn't so happy about this trick, but there's no nice way out of it anyway

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs (hunk 6 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs:487]: Just documented that we don't learn from those

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs [hunk 8]
```diff
@@ -433,6 +435,76 @@ public void Branching_LearnsObjectConstraint_TypePattern_Negated(string expressi
...
+            validator.ValidateTag("Else", x => x.HasConstraint(ObjectConstraint.Null).Should().BeTrue());
+            validator.TagValues("End").Should().HaveCount(2)
+                .And.ContainSingle(x => x.HasConstraint(ObjectConstraint.Null))
+                .And.ContainSingle(x => x.HasConstraint(ObjectConstraint.NotNull));
+        }
+
+        [DataTestMethod]
+        [DataRow("arg is string { }")]
+        [DataRow("arg is string { Length: 0 }")]
+        [DataRow("arg is string { Length: var length }")]
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs:458]:
```suggestion
        [DataRow("arg is string { Length: var length }")]
        [DataRow("arg is int { }")]
        [DataRow("arg is DateTime { Ticks: 0 }")]
        [DataRow("arg is TClass { }")]
        [DataRow("arg is T { }")]
        [DataRow("arg is TStruct { }")]
        [DataRow("arg is string { }", "TClass")]
        [DataRow("arg is string { }", "T")]        
```

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs [hunk 9]
```diff
@@ -433,6 +435,76 @@ public void Branching_LearnsObjectConstraint_TypePattern_Negated(string expressi
...
+            validator.ValidateTag("If", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue());
+            validator.ValidateTag("Else", x => x.Should().BeNull("it could be null or any other type"));
+            validator.TagValues("End").Should().HaveCount(2)
+                .And.ContainSingle(x => x == null)
+                .And.ContainSingle(x => x != null && x.HasConstraint(ObjectConstraint.NotNull));
+        }
+
+        [DataTestMethod]
+        [DataRow("arg is { }", "int")]
+        [DataRow("arg is { }", "T")]
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs:471]: In the true branch, we know that T is not null. Nullable analysis thinks so too:
image

https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4GIYFcA2OLA4CmABETAcQLABQAAgMwl0BMJAwiQN60l/NM6AFhIBZADwAVAHwAKSSQgAnAOYBKbr37aoAMxKzlKklADO3EgF81W7Xx407TxaoB0kgPYBlAC5LYKrJqANy2dpZhETSWQA
```suggestion
```

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.PatternMatching.cs [hunk 10]
```diff
@@ -345,19 +359,34 @@ public void DeclarationPattern_Var_Deconstruction_DoesNotSetConstraint()
...
         [DataRow("objectNotNull is int i", null)]
+        [DataRow("exceptionNull is object { }", false)]
+        [DataRow("exceptionNull is Exception { }", false)]
+        [DataRow("exceptionNull is FormatException { }", false)]
+        [DataRow("exceptionNotNull is object { }", true)]
+        [DataRow("exceptionNotNull is Exception { }", true)]
+        [DataRow("exceptionNotNull is FormatException { }", null)]
+        [DataRow("exceptionUnknown is object { }", null)]
+        [DataRow("exceptionUnknown is Exception { }", null)]
+        [DataRow("exceptionUnknown is FormatException { }", null)]
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.PatternMatching.cs:370]: Thanks!

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]
I added some more test cases (not sure if every test case is right, so take the suggestions with a grain of salt).

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs [hunk 11]
```diff
@@ -359,10 +360,11 @@ public void Branching_LearnsObjectConstraint_ConstantPattern_True(string express
...
         [DataTestMethod]
-        [DataRow("arg is not true")]
-        public void Branching_LearnsObjectConstraint_ConstantPattern_True_Negated(string expression)
+        [DataRow("arg is not true", "object")]
+        [DataRow("arg is not true", "bool")]
+        public void Branching_LearnsObjectConstraint_ConstantPattern_True_Negated(string expression, string argType)
         {
-            var validator = CreateIfElseEndValidatorCS(expression, OperationKind.ConstantPattern);
+            var validator = CreateIfElseEndValidatorCS(expression, OperationKind.ConstantPattern, argType);
             validator.ValidateTag("If", x => x.Should().BeNull("it could be False, null or any other type"));
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs:368]: This PR is for recursive patterns. I'll try to fix this after #6059 is merged

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs [hunk 12]
```diff
@@ -433,6 +435,76 @@ public void Branching_LearnsObjectConstraint_TypePattern_Negated(string expressi
...
+            validator.TagValues("End").Should().HaveCount(2)
+                .And.ContainSingle(x => x == null)
+                .And.ContainSingle(x => x != null && x.HasConstraint(ObjectConstraint.NotNull));
+        }
+
+        [DataTestMethod]
+        [DataRow("arg is { }", "int")]
+        [DataRow("arg is { }", "T")]
+        [DataRow("arg is { }", "TStruct")]
+        [DataRow("(arg, Unknown<object>()) is ({ }, { })", "object")]   // We don't support learning for tuples (yet). Should behave same as "arg is { }". Gets tricky when nesting (a, (b, c))
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs:473]: The argument is not optional on this one. As 3 out of 4 must have it, I didn't make it optional here

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs (hunk 8 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs:458]: Except `[DataRow("arg is string { }", "T")]` that didn't work (will not fix)

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs (hunk 9 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs:471]: T is not constraint and doesnt' have `IsReferenceType`. I won't fix it

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs (hunk 8 above)
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs:466]: What about `[DataRow("arg is string { }", "TClass")]`? It is missing too.

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]
Just one test case slipped through I think.
https://github.com/SonarSource/sonar-dotnet/pull/6058#discussion_r961702206

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs (hunk 8 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs:466]: Line 464
